score.go

```diff
@@ -0,0 +1,952 @@
+	MeshFailurePenaltyWeight, MeshFailurePenaltyDecay float64
+
+	// P4: invalid messages
+	// This is the number of invalid messages in the topic.
+	// The value of the parameter is a counter, decaying with InvalidMessageDeliveriesDecay.
+	// The weight of the parameter MUST be negative.
+	InvalidMessageDeliveriesWeight, InvalidMessageDeliveriesDecay float64
+}
+
+type peerStats struct {
+	// true if the peer is currently connected
+	connected bool
+
+	// expiration time of the score stats for disconnected peers
+	expire time.Time
+
+	// per topc stats
+	topics map[string]*topicStats
+
+	// IP tracking; store as string for easy processing
+	ips []string
+}
+
+type topicStats struct {
+	// true if the peer is in the mesh
+	inMesh bool
+
+	// time when the peer was (last) GRAFTed; valid only when in mesh
+	graftTime time.Time
+
+	// time in mesh (updated during refresh/decay to avoid calling gettimeofday on
+	// every score invocation)
+	meshTime time.Duration
+
+	// first message deliveries
+	firstMessageDeliveries float64
+
+	// mesh message deliveries
+	meshMessageDeliveries float64
+
+	// true if the peer has been enough time in the mesh to activate mess message deliveries
+	meshMessageDeliveriesActive bool
+
+	// sticky mesh rate failure penalty counter
+	meshFailurePenalty float64
+
+	// invalid message counter
+	invalidMessageDeliveries float64
+}
+
+type peerScore struct {
+	sync.Mutex
+
+	// the score parameters
+	params *PeerScoreParams
+
+	// per peer stats for score calculation
+	peerStats map[peer.ID]*peerStats
+
+	// IP colocation tracking
+	peerIPs map[string]map[peer.ID]struct{}
+
+	// message delivery tracking
+	deliveries *messageDeliveries
+
+	msgID MsgIdFunction
+	host  host.Host
+
+	// debugging inspection
+	inspect       PeerScoreInspectFn
+	inspectPeriod time.Duration
+}
+
+type messageDeliveries struct {
+	records map[string]*deliveryRecord
+
+	// queue for cleaning up old delivery records
+	head *deliveryEntry
+	tail *deliveryEntry
+}
+
+type deliveryRecord struct {
+	status    int
+	validated time.Time
+	peers     map[peer.ID]struct{}
+}
+
+type deliveryEntry struct {
+	id     string
+	expire time.Time
+	next   *deliveryEntry
+}
+
+// delivery record status
+const (
+	delivery_unknown   = iota // we don't know (yet) if the message is valid
+	delivery_valid            // we know the message is valid
+	delivery_invalid          // we know the message is invalid
+	delivery_throttled        // we can't tell if it is valid because validation throttled
+)
+
+type PeerScoreInspectFn func(map[peer.ID]float64)
+
+// WithPeerScoreDebug is a gossipsub router option that enables peer score debugging.
+// When this option is enabled, the supplied function will be invoked periodically to allow
+// the application to inspec or dump the scores for connected peers.
+// This option must be passed _after_ the WithPeerScore option.
+func WithPeerScoreDebug(inspect PeerScoreInspectFn, period time.Duration) Option {
```

Comment:
nit: homogeneous naming. `WithPeerScoreInspect`? Or `PeerScoreDumpFn`/`WithPeerScoreDump`?

Reply:
renamed to `WithPeerScoreInspect`.